File: DP1/DP1_1699_squareSum.cpp
#if 0
# Ive attended the code.plus class (https://code.plus/)
# And I've referred it's codes
# And core parts can be come fully from the class.
#endif


// d[n] = min num when representing sum with squarenums

// d[n] = min(d[n], d[n-i^2] + 1) , 1<=i , 0<= n-i^2
// d[0] = 0

// teacher
// d[n] = min(d[n-i^2] + 1),  1^2 <= i^2 <= N

#include <iostream>
#include <vector>
using namespace std;
int d[100001];

int main()
{
    int n;
    cin >> n;
    for (int i=1; i<=n; i++) {
        d[i] = i;
        // n * O(root n)
        for (int j=1; j*j <= i; j++) {
            if (d[i] > d[i-j*j] + 1) {
                d[i] = d[i-j*j]+1;
            }
        }
    }
    cout << d[n] << '\n';
    return 0;
} 

File: DP1/DP1_1699_squareSum.py
#if 0
# Ive attended the code.plus class (https://code.plus/)
# And I've referred it's codes
# And core parts can be come fully from the class.
#endif


n = int(input())
d = [0]*(n+1)
for i in range(1, n+1):
    d[i] = i
    j = 1
    while j*j <= i:
        if d[i] > d[i-j*j]+1:
            d[i] = d[i-j*j]+1
        j += 1
print(d[n])    
        